File: src/Tests/src/SchemaTests.cpp
#include <boost/filesystem.hpp>

#include "GTestGroups.h"

#include <Schema/schema.pa.h>

    using namespace boost;
    using namespace bsw;
    using namespace protoargs;

    class SchemaTests: public LogTest
    {
        public:
            SchemaTests() {}
            virtual ~SchemaTests() {}

        public:
            virtual void SetUp() override
            {
            }
    };

    TEST_F(SchemaTests, SchemaUsage)
    {
       const char* argv[] = {
          "program"
          ,"--help"
       };
       int argc = sizeof(argv)/sizeof(argv[0]);
       schema::ProtoArgs arguments;
       auto config = std::shared_ptr<schema::protoargs>( arguments.parse("program", argc, (char**)argv, true /*allow incomplete*/) );
       ASSERT_TRUE(config != nullptr);

       ASSERT_TRUE(config->printhelp());

       const auto usage = arguments.usage("program");
       printMessage( usage );
       ASSERT_FALSE( usage.empty() );
    }

    TEST_F(SchemaTests, CxxoptsConnectsShortAndLongArgsTogether)
    {
       const char* argv[] = {
          "program"
          , "-e", "valueE"
          , "--a-long-param", "somevalue" // inside pa.cc only '-a' is analized
          , "50" // paramG
          , "0" // bool paramG
          , "0.5" // PARAM-FLOAT
          , "0.7" // PARAM-DOUBLE
          , "pos1", "pos2", "pos3"
       };
       int argc = sizeof(argv)/sizeof(argv[0]);
       schema::ProtoArgs arguments;
       auto config = std::shared_ptr<schema::protoargs>( arguments.parse("program", argc, (char**)argv) );
       ASSERT_TRUE(config != nullptr);

       ASSERT_TRUE( config->has_parama() );
    }

    TEST_F(SchemaTests, SchemaPositiveShort)
    {
       const char* argv[] = {
          "program"
          ,"-e", "valueE"
          , "50" // paramG
          , "0" // bool paramG
          , "0.5" // PARAM-FLOAT
          , "0.7" // PARAM-DOUBLE
          , "pos1", "pos2", "pos3"
       };
       int argc = sizeof(argv)/sizeof(argv[0]);
       schema::ProtoArgs arguments;
       auto config = std::shared_ptr<schema::protoargs>( arguments.parse("program", argc, (char**)argv) );
       ASSERT_TRUE(config != nullptr);

       // check defaults
       ASSERT_FALSE( config->has_parama() );
       ASSERT_EQ( "// tricky default value", config->parama() );

       ASSERT_FALSE( config->has_paramb() );
       ASSERT_EQ( 10u, config->paramb() );

       ASSERT_FALSE( config->has_paramc() );
       ASSERT_EQ( 0, config->paramc() );

       ASSERT_FALSE( config->has_paramd() );
       ASSERT_EQ( 0.0f, config->paramd() );

       ASSERT_TRUE( config->has_parame() );
       ASSERT_EQ( "valueE", config->parame() );

       ASSERT_EQ( 0, config->paramf_size() );

       ASSERT_FALSE( config->has_param_i() );
       ASSERT_TRUE( config->param_i() );

       ASSERT_FALSE( config->has_param_j() );
       ASSERT_FALSE( config->param_j() );
    }

    TEST_F(SchemaTests, SchemaPositiveAll)
    {
       const char* argv[] = {
          "program"
          ,"-e", "valueE"
          ,"--a-long-param", "somevalue"
          ,"--b-long-param", "4"
          ,"-c", "555"
          ,"--d-long-param", "555.5"
          ,"-f", "1"
          ,"-f", "2"
          ,"-f", "3"
          ,"-i"
          ,"--j-long"
          , "50" // paramG
          , "0" // bool paramG
          , "0.5" // PARAM-FLOAT
          , "0.7" // PARAM-DOUBLE
          , "pos1", "pos2", "pos3" // paramH
       };
       int argc = sizeof(argv)/sizeof(argv[0]);
       schema::ProtoArgs arguments;
       auto config = std::shared_ptr<schema::protoargs>( arguments.parse("program", argc, (char**)argv) );
       ASSERT_TRUE(config != nullptr);

       ASSERT_TRUE( config->has_parama() );
       ASSERT_EQ( "somevalue", config->parama() );

       ASSERT_TRUE( config->has_paramb() );
       ASSERT_EQ( 4u, config->paramb() );

       ASSERT_TRUE( config->has_paramc() );
       ASSERT_EQ( 555, config->paramc() );

       ASSERT_TRUE( config->has_paramd() );
       ASSERT_EQ( 555.5f, config->paramd() );

       ASSERT_TRUE( config->has_parame() );
       ASSERT_EQ( "valueE", config->parame() );

       ASSERT_EQ( 3, config->paramf_size() );

       ASSERT_TRUE( config->has_param_i() );
       ASSERT_TRUE( config->param_i() );

       ASSERT_TRUE( config->has_param_j() );
       ASSERT_TRUE( config->param_j() );
    }

   /**
    * @brief Required means required
    */
    TEST_F(SchemaTests, MissingRequired)
    {
       const char* argv[] = {
          "program"
          , "50" // paramG
          , "0" // bool paramG
          , "pos1", "pos2", "pos3"
       };
       int argc = sizeof(argv)/sizeof(argv[0]);
       schema::ProtoArgs arguments;
       auto config = std::shared_ptr<schema::protoargs>( arguments.parse("program", argc, (char**)argv) );
       ASSERT_TRUE(config == nullptr);
    }

   /**
    * @brief Repeated positional args should ba at least one set
    */
    TEST_F(SchemaTests, MissingRepeatedPositional)
    {
       const char* argv[] = {
          "program"
          ,"-e", "valueE"
          , "50" // paramG
          , "0" // bool paramG
       };
       int argc = sizeof(argv)/sizeof(argv[0]);
       schema::ProtoArgs arguments;
       auto config = std::shared_ptr<schema::protoargs>( arguments.parse("program", argc, (char**)argv) );
       ASSERT_TRUE(config == nullptr);
    }

   /**
    * @brief Wrong type for positional arg
    */
    TEST_F(SchemaTests, PositionalWrongType)
    {
       const char* argv[] = {
          "program"
          ,"-e", "valueE"
          , "pos1", "pos2", "pos3"
       };
       int argc = sizeof(argv)/sizeof(argv[0]);
       schema::ProtoArgs arguments;
       auto config = std::shared_ptr<schema::protoargs>( arguments.parse("program", argc, (char**)argv) );
       ASSERT_TRUE(config == nullptr);
    }

File: src/TestsRust/src/schema_tests.rs
//use crate::simple_pa;
//#[path = "simple_pa.rs"] mod simple_pa;

#[cfg(test)]
mod schema_tests {

    // Note this useful idiom: importing names from outer (for mod tests) scope.
    //use super::*;

    #[test]
    fn schema_usage() {
        let description = "Desription";
        let argv = vec!["program", "--help"];

        // check usage
        let usage = crate::schema_pa::schema_pa::usage(argv[0], description);
        assert_eq!(usage.trim().is_empty(), false);

        // check configuration
        let rconfig = crate::schema_pa::schema_pa::parse_ext(argv[0], &argv[..], description, true);
        assert_eq!( true, rconfig.is_ok(), "{}", rconfig.err().unwrap() );
    }

    #[test]
    fn connect_short_and_long_args_together() {
        let description = "Desription";
        let argv = vec![ "program"
           , "-e", "valueE"
           , "--a-long-param", "somevalue" // inside only '-a' is analized
           , "50" // paramG
           , "true" // bool P-A-R-A-M-G-2
           , "0.5" // PARAM-FLOAT
           , "0.7" // PARAM-DOUBLE
           , "pos1", "pos2", "pos3"
        ];

        let rconfig = crate::schema_pa::schema_pa::parse_debug(argv[0], &argv[..], description);
        assert_eq!( true, rconfig.is_ok(), "{}", rconfig.err().unwrap() );
        let config = rconfig.ok().unwrap();

        assert!( config.has_parama() );
    }

    #[test]
    fn schema_positive_short() {
        let description = "Desription";
        let argv = vec![ "program"
           ,"-e", "valueE"
           , "50" // paramG
           , "false" // bool paramG
           , "0.5" // PARAM-FLOAT
           , "0.7" // PARAM-DOUBLE
           , "pos1", "pos2", "pos3"
        ];

        let rconfig = crate::schema_pa::schema_pa::parse_debug(argv[0], &argv[..], description);
        assert_eq!( true, rconfig.is_ok(), "{}", rconfig.err().unwrap() );
        let config = rconfig.ok().unwrap();

        // check defaults
        assert_eq!( false, config.has_parama() );
        assert_eq!( "// tricky default value", config.parama() );

        assert_eq!( false, config.has_paramb() );
        assert_eq!( 10, config.paramb() );

        assert_eq!( false, config.has_paramc() );
        assert_eq!( 0, config.paramc() );

        assert_eq!( false, config.has_paramd() );
        assert_eq!( 0.0, config.paramd() );

        assert_eq!( "valueE", config.parame() );

        assert_eq!( 0, config.paramf().len() );

        assert_eq!( false, config.has_param_i() );
        assert_eq!( true, config.param_i() );

        assert_eq!( false, config.has_param_j() );
        assert_eq!( false, config.param_j() );
    }

    #[test]
    fn schema_positive_all() {
        let description = "Desription";
        let argv = vec![ "program"
           ,"-e", "valueE"
           ,"--a-long-param", "somevalue"
           ,"--b-long-param", "4"
           ,"-c", "555"
           ,"--d-long-param", "555.5"
           ,"-f", "1"
           ,"-f", "2"
           ,"-f", "3"
           ,"-i"
           ,"--j-long"
           , "50" // paramG
           , "false" // bool paramG
           , "0.5" // PARAM-FLOAT
           , "0.7" // PARAM-DOUBLE
           , "pos1", "pos2", "pos3" // paramH
        ];

        let rconfig = crate::schema_pa::schema_pa::parse_debug(argv[0], &argv[..], description);
        assert_eq!( true, rconfig.is_ok(), "{}", rconfig.err().unwrap() );
        let config = rconfig.ok().unwrap();

        assert_eq!( true, config.has_parama() );
        assert_eq!( "somevalue", config.parama() );

        assert_eq!( true, config.has_paramb() );
        assert_eq!( 4, config.paramb() );

        assert_eq!( true, config.has_paramc() );
        assert_eq!( 555, config.paramc() );

        assert_eq!( true, config.has_paramd() );
        assert_eq!( 555.5, config.paramd() );

        assert_eq!( "valueE", config.parame() );

        assert_eq!( 3, config.paramf().len() );

        assert_eq!( true, config.has_param_i() );
        assert_eq!( true, config.param_i() );

        assert_eq!( true, config.has_param_j() );
        assert_eq!( true, config.param_j() );
    }

    #[test]
    //#[ignore] // FIXME Negative tests breaks execution
    fn missing_required() {
        let description = "Desription";
        let argv = vec![ "program"
           , "50" // paramG
           , "0" // bool paramG
        ];

        let rconfig = crate::schema_pa::schema_pa::parse_debug(argv[0], &argv[..], description);
        assert_eq!( false, rconfig.is_ok() );
    }

    #[test]
    //#[ignore] // FIXME Negative tests breaks execution
    fn missing_repeated_positional() {
        let description = "Desription";
        let argv = vec![ "program"
           ,"-e", "valueE"
           , "50" // paramG
           , "true" // bool paramG
           , "0.5" // PARAM-FLOAT
           , "0.7" // PARAM-DOUBLE
        ];

        let rconfig = crate::schema_pa::schema_pa::parse_debug(argv[0], &argv[..], description);
        assert_eq!( false, rconfig.is_ok() );
    }

    #[test]
    fn positional_wrong_type() {
        let description = "Desription";
        let argv = vec![ "program"
           ,"-e", "valueE"
           , "50f" // paramG
           , "0e" // bool paramG
           , "0.5d" // PARAM-FLOAT
           , "0.7d" // PARAM-DOUBLE
           , "pos1", "pos2", "pos3"
        ];

        let rconfig = crate::schema_pa::schema_pa::parse_debug(argv[0], &argv[..], description);
        assert_eq!( false, rconfig.is_ok() );
    }

}
